mathics/eval/assignment.py

```diff
@@ -0,0 +1,1461 @@
+# -*- coding: utf-8 -*-
+# pylint: disable-msg=too-many-arguments
+
+"""
+Support for Set and SetDelayed, and other assignment-like builtins
+"""
+
+from functools import reduce
+from typing import List, Optional, Tuple
+
+from mathics.core.assignment import (
+    build_rulopc,
+    get_symbol_list,
+    is_protected,
+    normalize_lhs,
+    rejected_because_protected,
+    unroll_conditions,
+    unroll_patterns,
+)
+from mathics.core.atoms import Atom, Integer, Integer1
+from mathics.core.attributes import A_LOCKED, attribute_string_to_number
+from mathics.core.builtin import Builtin
+from mathics.core.element import BaseElement
+from mathics.core.evaluation import (
+    MAX_RECURSION_DEPTH,
+    Evaluation,
+    set_python_recursion_limit,
+)
+from mathics.core.expression import Expression, SymbolDefault
+from mathics.core.rules import Rule
+from mathics.core.symbols import (
+    Symbol,
+    SymbolFalse,
+    SymbolMaxPrecision,
+    SymbolMinPrecision,
+    SymbolN,
+    SymbolTrue,
+    valid_context_name,
+)
+from mathics.core.systemsymbols import SymbolCondition, SymbolMachinePrecision
+from mathics.eval.list.eol import eval_Part
+
+
+class AssignmentException(Exception):
+    """Exception raised when Assignment fails"""
+
+    def __init__(self, lhs, rhs) -> None:
+        super().__init__(f" {rhs} cannot be assigned to {lhs}")
+        self.lhs = lhs
+        self.rhs = rhs
+
+
+class _SetOperator:
+    """
+
+    This is the base class for assignment Builtin operators.
```

Comment:
Stuff in eval probably should not have classes with objects. The focus of eval is an interpreter that knows nothing about OO. You should be able to write this code in C or assembly.

It is okay to pass OO objects or "self" pointers though. For these there are object _dereferences_,; Dereferencing is different from implicit OO method lookup which goes on inside classes. 

In sum, this class probably belongs in either `mathics.core` (similar to Atom classes) or `mathics.builtin`.

Reply:
Grepping `mathics.eval` right now I see: 

```console
$ PAGER=cat git grep -n "class "
distance/clusters.py:8:class IllegalDataPoint(Exception):
distance/clusters.py:12:class IllegalDistance(Exception):
files_io/read.py:42:class MathicsOpen(Stream):
patterns.py:11:class _StopGeneratorMatchQ(StopGenerator):
patterns.py:15:class Matcher:
pymathics.py:21:class PyMathicsLoadException(Exception):
rules.py:149:class Dispatch(Atom):
test.py:12:class _StopGeneratorBaseElementIsFree(StopGenerator):
```

Exceptions and Generators are weird kinds of classes, so there may be some wiggle room here. But `Dispatch(Atom)` and `Matcher` should probably be moved.

Reply:
The problem I found is that putting this class in another place produces circular dependencies among modules. I am trying to figure out how to avoid it.

Reply:
Easy: just get ride of the class...